### ISSUE TITLE
Speed up centerline extraction

### ISSUE THREAD

**Issue by mrp089** [SimVascular/SimVascular#963]
Introducing branch/junction splitting in #624 made centerline extraction much slower due to the need for slicing the surface at every centerline point (see #655). Centerline extraction depends on two steps: 1) generate a "raw" centerline in `VMTK`, 2) clean and split the centerline through a highly modified `VMTK CenterlineSections`. An easy speedup would be to significantly reduce the number of centerline points.

Currently, the number of centerline points generated by `VMTK` is quite excessive (several thousand). We don't need that level of detail for 0D/1D modeling or 3D result extraction. The sampling of the centerline in `VMTK` depends on mesh size, so a finer 3D mesh will have more centerline points. However, a "centerline mesh size" should probably depend on the actual model size rather than the 3D mesh size.

A solution could be to resample the centerline to a coarser resolution before running the branch/junction splitting. This could be done with this `VMTK` function: http://www.vmtk.org/vmtkscripts/vmtkcenterlineresampling.html

The resampling function would go between `sys_geom_centerlines` and `sys_geom_centerlinesections` in here: https://github.com/SimVascular/SimVascular/blob/7c07a4c9f95e34d45c19141591ac9b01371e64f3/Code/Source/sv4gui/Modules/Model/Common/sv4gui_ModelUtils.cxx#L1205-L1229

@zasexton is having a look at this. You can put any questions/problems that come up here.

**Comment by zasexton** [comment]
@mrp089 after looking at the vmtk repository I think the solution will be very straightforward. I don't think we will even need to add additional scripts to the simvascular_vmtk folder. "resampling" and "resamplingstep" are options during the [compute centerlines execution](http://www.vmtk.org/vmtkscripts/vmtkcenterlines.html). These should work well for downsampling the centerlines prior to slicing. I'll see if I can use those options today.

**Comment by zasexton** [comment]
Yes, the centerline resampling option is hard-coded to zero but included [here](https://github.com/SimVascular/SimVascular/blob/ab58b85aafdf29baa2a708767afca797d998ac59/Code/Source/sv/Mesh/VMTKUtils/sv_vmtk_utils.cxx).

**Comment by ktbolt** [comment]
Can't the centerline extraction be sped up by using isosurfaces to extract a slice?

**Comment by zasexton** [comment]
the procedure could definitely benefit from previous slicing extraction scripts that you've implemented for data curation. Since the centerline resampling was already an option I figured I'd expose it and see how well it works. But ultimately I think it's a quick fix but something more robust, like isosuface extraction, would probably be better

—
Stanford Bioengineering
NSF Graduate Fellow
________________________________

**Comment by zasexton** [comment]
I've made changes to the following three files to expose centerline resampling options for vmtk centerlines. The links display the original code and the code foollowing them are my associated changes. I've only added these for the PythonAPI and not the GUI. 

- SimVascular/Code/Source/sv/Mesh/VMTKUtils/sv_vmtk_utils.cxx
- SimVascular/Code/Source/sv/Mesh/VMTKUtils/sv_vmtk_utils.h
- SimVascular/Code/Source/PythonAPI/Vmtk_PyModule.cxx

First I included the resample and resampling step arguments in the sys_geom_centerlines function and removed the previously hard-coded values. 
https://github.com/SimVascular/SimVascular/blob/158eb1ce28ed92df051650612ce0b97dcb6e1658/Code/Source/sv/Mesh/VMTKUtils/sv_vmtk_utils.cxx#L112-L166 

```c++
int sys_geom_centerlines( cvPolyData *polydata,int *sources,int nsources,
                int *targets,int ntargets,
                cvPolyData **lines, cvPolyData **voronoi, int resample,
                float resamplingstep)
{
  vtkPolyData *geom = polydata->GetVtkPolyData();
  cvPolyData *result1 = NULL;
  cvPolyData *result2 = NULL;
  *lines = NULL;
  *voronoi = NULL;

//  vtkSmartPointer<vtkvmtkPolyDataCenterlines> centerLiner =
//    vtkSmartPointer<vtkvmtkPolyDataCenterlines>::New();
  vtkSmartPointer<vtkIdList> capInletIds = vtkSmartPointer<vtkIdList>::New();
  vtkSmartPointer<vtkIdList> capOutletIds = vtkSmartPointer<vtkIdList>::New();
  int pointId;

  fprintf(stderr,"NumSources %d\n",nsources);
  fprintf(stderr,"NumTargets %d\n",ntargets);

  for (pointId = 0;pointId<nsources;pointId++)
  {
    capInletIds->InsertNextId(sources[pointId]);
  }
  for (pointId = 0;pointId<ntargets;pointId++)
  {
    capOutletIds->InsertNextId(targets[pointId]);
  }
  vtkNew(vtkvmtkPolyDataCenterlines,centerLiner);
  try {
    //std::cout<<"Getting Center Lines..."<<endl;
    centerLiner->SetInputData(geom);
    centerLiner->SetSourceSeedIds(capInletIds);
    centerLiner->SetTargetSeedIds(capOutletIds);
    centerLiner->SetRadiusArrayName("MaximumInscribedSphereRadius");
    centerLiner->SetCostFunction("1/R");
    centerLiner->SetFlipNormals(0);
    centerLiner->SetAppendEndPointsToCenterlines(1);
    centerLiner->SetSimplifyVoronoi(0);
    centerLiner->SetCenterlineResampling(resample);
    centerLiner->SetResamplingStepLength(resamplingstep);
    centerLiner->Update();

    result1 = new cvPolyData( centerLiner->GetOutput() );
    *lines = result1;
    result2 = new cvPolyData( centerLiner->GetVoronoiDiagram() );
    *voronoi = result2;
  }
  catch (...) {
    fprintf(stderr,"ERROR in centerline operation.\n");
    fflush(stderr);
    return SV_ERROR;
  }

  return SV_OK;
}
```

I then included the optional arguments in the associated header file.

https://github.com/SimVascular/SimVascular/blob/158eb1ce28ed92df051650612ce0b97dcb6e1658/Code/Source/sv/Mesh/VMTKUtils/sv_vmtk_utils.h#L54-L56

```c++
SV_EXPORT_VMTK_UTILS int sys_geom_centerlines( cvPolyData *polydata, int *source, int nsources,
                            int *targets, int ntargets,
                            cvPolyData **lines, cvPolyData **voronoi, int resampling = 0,
                            float resamplingstep=1);

```

I then updated the Vmtk_PyModule to reflect these changes for the python interface.
https://github.com/SimVascular/SimVascular/blob/158eb1ce28ed92df051650612ce0b97dcb6e1658/Code/Source/PythonAPI/Vmtk_PyModule.cxx#L349-L502

```c++
//------------------
// Vmtk_centerlines
//------------------
//
PyDoc_STRVAR(Vmtk_centerlines_doc,
   "centerlines(surface, inlet_ids, outlet_ids, split=True,                \n\
        use_face_ids=False, resample=False, resample_step=1)               \n\
   \n\
   Compute the centerlines for a closed surface.                            \n\
   \n\
   Args: \n\
     surface (vtkPolyData): The vtkPolyData object representing a closed   \n\
        surface. \n\
     inlet_ids (list[int]): The list of integer IDs identifying the vessel \n\
        inlet faces.                                                       \n\
     outlet_ids (list[int]): The list of integer IDs identifying the vessel\n\
        outlet faces. \n\
     split (bool): If True then split centerlines into branches.           \n\
        they are node IDs.                                                 \n\
     use_face_ids (bool): If True then the input IDs are face IDs, else    \n\
        they are node IDs.                                                 \n\
     resample (bool): If True then the centerline will be resampled according to \n\
               the resampling step parameter.                             \n\
     resample_step (float): The step length defining the resampling size for\n\
                            a given centerline.                            \n\
   \n\
   Returns (vtkPolyData): The centerlines geometry (lines) and data.       \n\
");

static PyObject *
Vmtk_centerlines(PyObject* self, PyObject* args, PyObject* kwargs)
{
  //std::cout << "========== Vmtk_centerlines ==========" << std::endl;
  auto api = PyUtilApiFunction("OO!O!|O!O!O!d", PyRunTimeErr, __func__);
  static char *keywords[] = {"surface", "inlet_ids", "outlet_ids", "split", "use_face_ids", "resample", "resample_step", NULL};
  PyObject* surfaceArg;
  PyObject* inletIdsArg;
  PyObject* outletIdsArg;
  PyObject* splitArg = nullptr;
  bool splitCenterlines = true;
  PyObject* useFaceIdsArg = nullptr;
  bool useFaceIds = false;
  PyObject* useResample = nullptr;
  int resample = 0;
  double resamplestep = 1.0;

  if (!PyArg_ParseTupleAndKeywords(args, kwargs, api.format, keywords, &surfaceArg, &PyList_Type, &inletIdsArg, 
        &PyList_Type, &outletIdsArg, &PyBool_Type, &splitArg, &PyBool_Type, &useFaceIdsArg, &PyBool_Type, &useResample, &resamplestep)) {
      return api.argsError();
  }

  // Check inlet IDs.
  //
  // [TODO:DaveP] Add this as a util function.
  //
  std::vector<int> sources;
  int numInletIds = PyList_Size(inletIdsArg);
  if (numInletIds == 0) {
    api.error("The 'inlet_ids' argument is empty.");
    return nullptr;
  }
  for (int i = 0; i < numInletIds; i++) {
      auto item = PyList_GetItem(inletIdsArg, i);
      if (!PyLong_Check(item)) {
          api.error("The 'inlet_ids' argument is not a list of integers.");
          return nullptr;
      }
      int id = PyLong_AsLong(item);
      //std::cout << "[Vmtk_centerlines]   ID: " << id << std::endl;
      sources.push_back(id);
  }

  // Get the target IDs.
  //
  int numOutletIds = PyList_Size(outletIdsArg);
  std::vector<int> targets;
  if (numOutletIds == 0) {
    api.error("The 'outlet_ids' argument is empty.");
    return nullptr;
  }
  for (int i = 0; i < numOutletIds; i++) {
      auto item = PyList_GetItem(outletIdsArg, i);
      if (!PyLong_Check(item)) {
          api.error("The 'outlet_ids' argument is not a list of integers.");
          return nullptr;
      }
      int id = PyLong_AsLong(item);
      targets.push_back(id);
  }
  // Check for IDs given as both sources and targets.
  //
  std::vector<int> commonIds;
  std::set_intersection(sources.begin(), sources.end(), targets.begin(), targets.end(), std::back_inserter(commonIds));
  if (commonIds.size() != 0) {
      std::ostringstream ids;
      std::copy(commonIds.begin(), commonIds.end()-1, std::ostream_iterator<int>(ids, ", "));
      ids << commonIds.back();
      api.error("The 'inlet_ids' and 'outlet_ids' arguments contain identical IDs '" + ids.str() + "'.");
      return nullptr;
  }

  // Get the vtkPolyData object from the Python object.
  //
  auto surfPolydata = GetVtkPolyData(api, surfaceArg);
  if (surfPolydata == nullptr) {
      return nullptr;
  }
  cvPolyData cvSurfPolydata(surfPolydata);

  // If use face IDs then map the face IDs for
  // the source and targets to node IDs.
  //
  if (useFaceIdsArg != nullptr) {
      useFaceIds = PyObject_IsTrue(useFaceIdsArg);
  }
  if (useFaceIds) {
      sources = ConvertFaceIdsToNodeIds(api, surfPolydata, sources);
      targets = ConvertFaceIdsToNodeIds(api, surfPolydata, targets);
      if ((sources.size() == 0) || (targets.size() == 0)) {
          return nullptr;
      }
  }
  if (useResample != nullptr){
      useResample = PyObject_IsTrue(useResample);
      if (useResample) {
          resample = 1;
      }
  }

  // Calculate the centerlines.
  cvPolyData* linesDst = nullptr;
  cvPolyData* voronoiDst = nullptr;
  //std::cout << "[Vmtk_centerlines] Calculate centerlines ... " << std::endl;

  if (sys_geom_centerlines(&cvSurfPolydata, sources.data(), numInletIds, targets.data(), numOutletIds, &linesDst, &voronoiDst, useResampl$
      api.error("Error calculating centerlines.");
      return nullptr;
  }

  // If split centerlines into branches. 
  //
  if (splitArg != nullptr) {
      splitCenterlines = PyObject_IsTrue(splitArg);
  }

  if (splitCenterlines) {
      cvPolyData* splitCenterlines = nullptr;
      cvPolyData* surfGrouped = nullptr;
      cvPolyData* sections = nullptr;
      if (sys_geom_centerlinesections(linesDst, &cvSurfPolydata, &splitCenterlines, &surfGrouped, &sections) != SV_OK) {
          api.error("Error splitting centerlines.");
          delete linesDst;
          delete surfGrouped;
          delete sections;
          return nullptr;
      }
      delete surfGrouped;
      delete sections;
      delete linesDst;
      linesDst = splitCenterlines;
  }
  //std::cout << "[Vmtk_centerlines] Done. " << std::endl;

  return vtkPythonUtil::GetObjectFromPointer(linesDst->GetVtkPolyData());
}

```
Let me know if I've done anything outrageously egregious with the code. I'm also including three images below. The first is the original centerline with 23019 slicing points followed by resampled centerlines with 1247 and 124 points, respectively.
![cl_34803](https://user-images.githubusercontent.com/47196674/159332479-a51f7405-e60b-4703-9191-ce5b403c7039.png)
![cl_1247](https://user-images.githubusercontent.com/47196674/159332523-36b448a9-12c7-4a41-a602-21dd0bb46d70.png)
![cl_124](https://user-images.githubusercontent.com/47196674/159332529-c378932f-61f1-44b1-8f96-c6a7bfc55da5.png)

**Comment by mrp089** [comment]
Awesome work @zasexton!! I think this is a very elegant solution - use resampling functionality that's already provided in the centerline extraction and set default values to "no resampling".

Can you add your changes to a [feature branch](https://www.atlassian.com/git/tutorials/comparing-workflows/feature-branch-workflow)? Then we can look at the code and run it as well (e.g. test it on all VMR models).

Also, for your examples, what parameters did you use (this will be a challenge to determine automatically)? Do you have the branch/junction splitting for your centerlines as well (would be interesting to see how the resampling affects that)? If so, can you upload them so we can look at them in 3D?

@ktbolt I agree that a more elegant slicing solution as implemented by you and @lucapegolotti would speed up things as well. But in the end, it also scales linearly with the number of centerline points. Since the number of points can be excessively large (much larger than we need for any application) I think this is a good starting point to reduce computational effort.

Even if we don't "ship" SimVascular with automatic resampling of centerlines, I think it's still good to expose that option more in the code (for developers). Instead of automatic resampling, we could also offer an option to the users to resample the centerline (with an automatically computed default value).

In addition, the resampling might have a positive side-effect on branch/junction splitting. The transition between branches in the resampled versions looks more like how blood vessels are connected and less like streamlines of fluid particles. That might allow for smaller junction regions which could break up some of the "monster junctions" (>20 outlets) in pulmonary models into smaller junctions with fewer outlets. But more testing is required for that...

**Comment by ktbolt** [comment]
@zasexton In general this looks fine to me. A couple of comments

1) Don't return an int into `useResample`, it is a `PyObject*`

```
if (useResample != nullptr){
      useResample = PyObject_IsTrue(useResample);
      if (useResample) {
          resample = 1;
      }
  }
```

just check it
```
if (useResample != nullptr){
      if (PyObject_IsTrue(useResample)) {
          resample = 1;
      }
  }
```

2) It is best to place returned function parameters at the end of the function definition. For example, replace
```
int sys_geom_centerlines( cvPolyData *polydata,int *sources,int nsources,
                int *targets,int ntargets,
                cvPolyData **lines, cvPolyData **voronoi, int resample,
                float resamplingstep)
```
with
```
int sys_geom_centerlines( cvPolyData *polydata,int *sources, int nsources,
                int *targets,int ntargets, int resample,
                float resamplingstep, cvPolyData **lines, cvPolyData **voronoi, 
```

And it would be even better to identify parameters using `const` like so 
```
int sys_geom_centerlines( const cvPolyData *polydata, const int *sources, const int nsources, const int *targets, const
    int ntargets, cvPolyData **lines, cvPolyData **voronoi)
```
but SV does not typically do this and methods are not declared `const`. 


3) What happens if `resamplingstep` (rename this to `resamplingStep`) is a value that does not make sense, too large or too small? Does the exception return any useful information? Is it possible to check this value in the Python API?

4) Add comments to `sys_geom_centerlines`.

**Comment by zasexton** [comment]
I've included these changes and incorporated most of the comments here https://github.com/zasexton/SimVascular/commit/77957c048deb03ed27c1212b8caedeab931323a0. @ktbolt when moving the `resample` and `resamplingStep` arguments before `lines` and `voronoi` there were errors when compiling because I assign them default arguments in the header file. This current commit works as is, but is there a more elegant way to assign default arguments outside the header so that the parameter order can have the returned function parameters at the end?

**Comment by ktbolt** [comment]
@zasexton It is too bad `sys_geom_centerlines` passes data back using parameters rather than using a return, you will need to set defaults for the remaining parameters to get `sys_geom_centerlines` to compile like this
```
int sys_geom_centerlines( cvPolyData *polydata, int *sources,int nsources, int *targets, int ntargets,  
    bool resample = false, float resamplingStep = 0.5, 
    cvPolyData **lines = nullptr, cvPolyData **voronoi = nullptr)		
```

The `resample` parameter should be declared a `bool` data type, makes it clearer what this parameter is used for, change that for the Python API too.

**Comment by zasexton** [comment]
@ktbolt Oop! Thanks for catching that. I've made that change now here https://github.com/zasexton/SimVascular/commit/9c4ea1b9a85a9b42e9148eda3267c214cef2fc5d. @mrp089 I've included a zip folder containing three centerlines resampled three different resampling steps. The parameter values cooresponding to the respective centerline are given in the readme.txt. The original 3D model is included as well as the inlet and outlet face ids needed to run the vmtk centerline command from the PythonAPI. 
[centerline_resampling_test.zip](https://github.com/SimVascular/SimVascular/files/8327951/centerline_resampling_test.zip)

**Comment by mrp089** [comment]
Thanks for the examples, @zasexton! Looking at the results, I discovered a problem with this approach.

Currently, the resampling works on individual centerlines, i.e., a connection from the inlet to one of the outlets. These centerlines are then merged in [vtkvmtkPolyDataCenterlineSections.cxx](https://github.com/SimVascular/SimVascular/blob/master/Code/ThirdParty/vmtk/simvascular_vmtk/vtkvmtkPolyDataCenterlineSections.cxx). Thus, we do the resampling **before** merging the centerlines. I zoomed in on the geometry you uploaded. Left is no resampling, right is resampling_step = 0.1 (coloring by `BranchIdTmp`):
![centerlines_resampling](https://user-images.githubusercontent.com/51269309/159589736-4a86b607-cd62-449e-b9af-770bfd78cb57.png)

If centerlines are resampled individually, it looks like they change their "path" and not just their resolution. Thus, the points that should be merged are no longer merged and thus detected as individual branches. Therefore, I think the resampling has to happen **after** the merge. 

The merging happens here:
https://github.com/SimVascular/SimVascular/blob/7c07a4c9f95e34d45c19141591ac9b01371e64f3/Code/ThirdParty/vmtk/simvascular_vmtk/vtkvmtkPolyDataCenterlineSections.cxx#L1044-L1050

@zasexton Do you think you can call `ResampleCenterlines` in [vtkvmtkPolyDataCenterlines.cxx](https://github.com/SimVascular/SimVascular/blob/master/Code/ThirdParty/vmtk/simvascular_vmtk/vtkvmtkPolyDataCenterlines.cxx) directly after with the cleaned centerline? (Alternatively, we would have to come up with a different resampling technique that doesn't move points but just drops them.)

**Comment by zasexton** [comment]
It might be difficult to call `ResampleCenterlines` from the PythonAPI. As I understand only polydata objects are returned so I don't believe we'd be able to call this method after the fact. I'll look into this now and see.

**Comment by zasexton** [comment]
@mrp089 do you call [`sys_geom_mergecenterlines`](https://github.com/SimVascular/SimVascular/blob/7c07a4c9f95e34d45c19141591ac9b01371e64f3/Code/Source/sv/Mesh/VMTKUtils/sv_vmtk_utils.cxx#L190-L221) ? From my understanding the [`sys_geom_centerlinesection`](https://github.com/SimVascular/SimVascular/blob/7c07a4c9f95e34d45c19141591ac9b01371e64f3/Code/Source/sv/Mesh/VMTKUtils/sv_vmtk_utils.cxx#L271-L342) relies exclusively on[ `vtkvmtkPolyDataCenterlineSections`](https://github.com/SimVascular/SimVascular/blob/master/Code/ThirdParty/vmtk/simvascular_vmtk/vtkvmtkPolyDataCenterlineSections.cxx) to split and slice. If so any resampling will probably have to happen within this function.

**Comment by mrp089** [comment]
@zasexton I see what the problem is, `ResampleCenterlines` works directly on the `vtkSVCenterlines` object and can thus not be called from the outside with an arbitrary `vtkPolyData` centerline.

Maybe one way could be to c&p  `ResampleCenterlines` into `vtkvmtkPolyDataCenterlineSections.cxx` and call it somewhere in `GenerateCleanCenterline`.

**Comment by zasexton** [comment]
That's what I have begun working on. I might have to talk with you more about how to implement this exactly. I'm having errors after `GeneratingCleanCenterline` and in `RefineCapPoints`

**Comment by zasexton** [comment]
This version of resampling should better handle centerline resampling https://github.com/zasexton/SimVascular/commit/b8d80d245f53f2df448b38e2b87c1126ccc4c311 I still have to clean up a lot around it and will do so tomorrow to hopefully finish most of this issue. I will include the test cases and rationale in that followup comment. Just wanted to include this here for progress.

**Comment by zasexton** [comment]
The most current version of centerline resampling https://github.com/zasexton/SimVascular/commit/0e6ab9bcca6a3ac2f00e01dd07e7e0d65e4eb297 have asked @mrp089 to review centerlineID assignment. I keep having occasional endpoints included in other centerline Ids. Examples of this can be seen in the attached data.
[centerline_files.zip](https://github.com/SimVascular/SimVascular/files/8374031/centerline_files.zip)

**Comment by zasexton** [comment]
Found the bug and fixed the mislabeled centerlines at the endpoints for resampled centerlines the reflected changes are included here https://github.com/zasexton/SimVascular/commit/76094a19c585d0e24a06bf4284097e34175078e2. I've also included a zip file of a test case with the solid model, original (non-resampled) centerlines, and resampled centerlines for comparison. @mrp089 it might be nice to also test this on the vascular model repository as you mentioned before to make sure that this method will hold up.
[centerline_files.zip](https://github.com/SimVascular/SimVascular/files/8415952/centerline_files.zip)

**Comment by mrp089** [comment]
Awesome, @zasexton! I can run your code with all VMR models. What `ResampleStep` should I use?

**Comment by zasexton** [comment]
@mrp089 Thank you so much for running this! I'll go through and compare. What `resampling_step` did you use? Also would it be possible to teach me how you ran this in case I wanted to include the speedup?

**Comment by zasexton** [comment]
I've included a few slides here of what I noticed.
[centerline resampling results.pptx](https://github.com/SimVascular/SimVascular/files/8429521/centerline.resampling.results.pptx)
The approach might need to be more careful when downsampling highly curved vessels.

**Comment by mrp089** [comment]
Thank you for the great summary, @zasexton! Maybe there is a way to treat this as an "optimization problem": given the number of points `N`, how can we position the points so we best approximate the centerline?

[Here](https://github.com/StanfordCBCL/DataCuration/blob/master/get_centerline.py) is the code I used to generate the centerlines and the VMR [surfaces](https://drive.google.com/file/d/1FpV1S-7DeDeuS2Y885anz7LxRKTFZXIr/view?usp=sharing) and [refined surfaces](https://drive.google.com/file/d/1_u_n4WWsImdAHFU_2RqlQvK4LNyBHmyH/view?usp=sharing) for some models.

**Comment by zasexton** [comment]
@mrp089 I know you've been working on some of the 1D simulations recently. Have you given any more thoughts to centerline extraction speedup? I thought that the solutions (more like hacks) that I've been implementing were okay but not robust enough to justify incorporating them into SV yet. if our wise c++ guru @ktbolt has any thoughts on this it'd be nice to get your perspective as well

**Comment by ktbolt** [comment]
I would profile the code and see what's taking up the majority of the compute time. There are three primary components I think
1) Computing the Voronoi diagram - might be O(n log(n)) or O(n*n) depending on the algorithm
2) Solving the Eikonal equation - O(n log(n))
3) Backtracking - not sure

You could then look at optimizing the code: maybe replace some VTK calls or possibly add some thread-based parallelism. 

Doesn't the slice extract time dwarf the centerline compute time?

**Comment by mrp089** [comment]
@ktbolt Just remembered this issue after our conversation yesterday

**Comment by ktbolt** [comment]
@mrp089 Yes, I spoke with @zasexton about this yesterday. Maybe it is time to decide what to do.

**Comment by zasexton** [comment]
Yes, when there is time I have a few ideas written down from when I tried some fixes originally. Maybe me and @mrp089 can meet to talk about some of these to come up with a concrete plan for implementing a true fix

**Comment by spaceie08** [comment]
did you guys fix it?
